Friends app by Evgen241984

Evgeniy241984 opened the pull request:
### Friends app 

[Demo](https://evgeniy241984.github.io/friends_app/)
[Code base](https://github.com/Evgeniy241984/frontend-2021-homeworks/blob/friends-app/submissions/evgeniy24/friends_app/app.js)

The code is submitted in a dedicated feature branch.
Please, review.

zonzujiro commented:
I see some common mistakes - pls check this AMA and fix code using information from it https://www.youtube.com/watch?v=iqdUtYGYdDE

Also, this comment https://github.com/kottans/frontend-2021-homeworks/pull/425#discussion_r560993334 applies to your code also

Evgeniy241984 commented:
> I see some common mistakes - pls check this AMA and fix code using information from it https://www.youtube.com/watch?v=iqdUtYGYdDE
> 
> Also, this comment [#425 (comment)](https://github.com/kottans/frontend-2021-homeworks/pull/425#discussion_r560993334) applies to your code also

Thank you for comments! 
I refactored variables/function as it was said in the AMA! Also refactored sortName/sortAge functions according #425 (comment) 
Pray for your health!  Please see my code.

Evgeniy241984 commented:
> > I see some common mistakes - pls check this AMA and fix code using information from it https://www.youtube.com/watch?v=iqdUtYGYdDE
> > Also, this comment [#425 (comment)](https://github.com/kottans/frontend-2021-homeworks/pull/425#discussion_r560993334) applies to your code also
> 
> Thank you for the comments!
> I refactored variables/function as it was said in the AMA,
> added error handling to the fetch method,
> refactored sortName/sortAge functions according to #425 (comment).
> Pray for your health! Please see my code.

After your ansfer at yesterday's  AMA  -  i'm refactoring the code!

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,234 @@
+    genderNeutralList, filteredByGenderList,
+    scrollDepth;
+    
+const sendRequest = function(method, url) {
```
`method` not used

Evgeniy241984 replied:
Yes! Deleted!

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,234 @@
+
+startApp();
+
+// drawing a list of friends---------------------------
```
Redundant comment. Please, remove all of them

Evgeniy241984 replied:
All comments have been removed!

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,234 @@
+    const mainInfo = document.createElement('div');
+    const friendsName = document.createElement('a');
+    const moreWrap = document.createElement('div');
+    const friendPhotoWrap = document.createElement('div')
+    const friendPhoto = document.createElement('img');
+    const friendAge = document.createElement('p');
+    const friendEmail = document.createElement('a');
+    const friendPhone = document.createElement('a');
+    const friendCity = document.createElement('a');
+
+    cardWrap.classList.add('cardWrap');
+    (friend.gender === 'male') ? cardWrap.classList.add('malegradient') : cardWrap.classList.add('femalegradient') 
+    mainInfo.classList.add('mainInfo');
+    moreWrap.classList.add('moreWrap');
+    friendPhotoWrap.classList.add('cardWrap__photo-wrap');
+    friendPhoto.classList.add('cardWrap__photo-img');
+    friendsName.classList.add('mainInfo__friendsName');
+    friendAge.classList.add('mainInfo__age');
+    friendEmail.classList.add('moreWrap__email');
+    friendPhone.classList.add('moreWrap__phone');
+    friendCity.classList.add('mainInfo__city');
+    
+    friendsName.textContent = `${friend.name.first} ${friend.name.last}`;
+    friendAge.textContent = `Age: ${friend.dob.age}`;
+    friendEmail.textContent = `email: ${friend.email}`;
+    friendPhone.textContent = `phone: ${friend.phone}`;
+    friendCity.textContent = `City: ${friend.location.city}`;
+
+    friendPhoto.setAttribute('src', friend.picture.medium);
+    friendsName.setAttribute('href', `https://www.facebook.com/search/top?q=${friend.name.first}%20${friend.name.last}`);
+    friendsName.setAttribute('target', '_blank');
+    friendEmail.setAttribute('href', `mailto:${friend.email}`);
+    friendPhone.setAttribute('href', `tel:>${friend.phone}`);
+    friendCity.setAttribute('href', `https://www.google.com/maps/place/${friend.location.city}$`);
+    friendCity.setAttribute('target', '_blank');
+
+    moreWrap.append(friendPhone, friendEmail);
+    mainInfo.append(friendsName, friendAge, friendCity);
+    friendPhotoWrap.append(friendPhoto);
+    cardWrap.append(friendPhotoWrap, mainInfo, moreWrap);
```
Just replace it with template string and `innerHtml`

Evgeniy241984 replied:
Ok! Done!

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,234 @@
+        sortedList.sort((a, b) => sortByName(b, a));
+        renderFriendsList(sortedList);
+
+    } else if (event.target === SORT_BY_AGE_YO) {
```
what the meaning of `YO`? 🤔 It's a ninja code ;)

Evgeniy241984 replied:
Young Old and vice versa!)
I will rename it

Evgeniy241984 replied:
Renamed! See a new commit!

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,234 @@
+        renderFriendsList(sortedList);
+
+    } else if (event.target === SORT_BY_AGE_YO) {
+        sortedList.sort((a, b) => sortByAge(a, b));
```
```suggestion
        sortedList.sort(sortByAge);
```

Evgeniy241984 replied:
Refactored!

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,234 @@
+})
+
+//filtering list of friends by gender
+FILTER_MENU.addEventListener('click', function(event) {
```
Your checkbox can have value `male`, `female`, etc.  Same with age and name

Then you can just filter by value of input, without redundant checks on `target`.

Evgeniy241984 replied:
Yep! Refactored!

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    const sortByAge = function(a, b) {
+        return (a.dob.age - b.dob.age);
```
Brackets redundant.

Please, use Prettier :)

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,234 @@
+
+    } else if (event.target === FILTER_BY_ALL) {
+        filteredByGenderList = [...friendsList];
+        renderFriendsList(filteredByGenderList);
```
Same as at lines 168, 174. DRY :)

It can be called once after all operations

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,234 @@
+    } else if (event.target === FILTER_BY_ALL) {
+        filteredByGenderList = [...friendsList];
+        renderFriendsList(filteredByGenderList);
+        }
```
Redundant spacing. Use prettier :)

Evgeniy241984 replied:
I thought I was using. 
it was installed but not activated ))

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,234 @@
+
+// showing user message when data loading error ------------------------------------------------------
+function renderErrorMessage() {
+    const messageErrorWrap =  document.createElement('div');
+    const messageError = document.createElement('p');
+    const tryMoreBtn = document.createElement('button');
+
+    messageErrorWrap.classList.add('error-message-wrap')
+    tryMoreBtn.classList.add('btn');
+    tryMoreBtn.classList.add('error-btn');
+    messageError.classList.add('error-msg-text');
+
+    messageError.textContent = 'OOPS! Something Bad Happened, please try again';
+    tryMoreBtn.textContent = 'Try Again';
+
+
+    messageErrorWrap.append(messageError, tryMoreBtn);
```
Use template string

Evgeniy241984 replied:
Refactored ! See a new commit!

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,234 @@
+
+// when error - reSend Request to server -----------------------
+FRIENDS_LIST_WRAP.addEventListener('click', function(event) {
+    if (event.target.classList == 'btn error-btn') {
```
What if classname for error will be changed? 🤔

Evgeniy241984 replied:
Done, refactored ! See a new commit!

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,234 @@
+    MENU_WRAP = document.querySelector('.menu-wrap'),
+    SORTING_MENU = document.querySelector('.filter-sort-menu'),
+    SEARCH_FIELD = document.querySelector('#search'),
+    SORT_BY_NAME_AZ = document.querySelector('.nameSortAZ'),
+    SORT_BY_NAME_ZA = document.querySelector('.nameSortZA'),
+    SORT_BY_AGE_YO = document.querySelector('.ageSortYoungerFirst'),
+    SORT_BY_AGE_OY = document.querySelector('.ageSortOlderFirst'),
```
Redundant because you can use `value` from checkboxes

Evgeniy241984 replied:
Done, refactored !

zonzujiro commented on submissions/evgeniy24/friends_app/random-color-header.js:
```diff
@@ -0,0 +1,36 @@
+    )
+}
+
+Array.prototype.randomColor = function() {
```
Eeehm... Why it should be in prototype?

And in separate file?

Evgeniy241984 replied:
I found this solution on the Internet, I liked it and decided to implement it in my friends app.
To make it easier to work with the code during implementation, it worked in a separate file!
In a new commit, I moved everything to the main file.

zonzujiro replied:
It's a bad solution :)

First of all, you [monkey patched](https://en.wikipedia.org/wiki/Monkey_patch) the prototype of the array. Check part with risks.

Also, when you do such things, you should check case if a method with your name exists in the prototype, just not to override things. Imagine you have added a method called `filter` and then JS will also ship such a method. You will override existing functionality. Check story about [polyfills](https://developer.mozilla.org/en-US/docs/Glossary/Polyfill)

Aaaand... copy-pasting solutions from internet leads to [SODD](https://dzone.com/articles/stack-overflow-driven-development-sodd-its-really)

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,248 @@
+    scrollDepth,
+    lettersArr = HEADER.innerHTML.split('');
+
+//convert int to hex
```
Redundant comment

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,248 @@
+    const FRIEND_CARD = document.createElement('div');
+    FRIEND_CARD.innerHTML =
+        `
+    <div class="cardWrap ${friend.gender === 'male' ? 'malegradient' : 'femalegradient'}">
```
You have invented two new words? `malegradient` and second one :)

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,248 @@
+        return a.dob.age - b.dob.age;
+    }
+
+    if (event.target.value === 'az') {
```
`az`, `za`, `abc` :) Common :D Find a better naming

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,248 @@
+        sortedList = [...friendsList];
+    }
+
+    const searchByName = function (a, b) {
+        const nameA = a.name.first.toUpperCase();
+        const nameB = b.name.first.toUpperCase();
+        return nameA < nameB ? -1 : nameA > nameB ? 1 : 0;
+    }
+
+    const sortByAge = function (a, b) {
+        return a.dob.age - b.dob.age;
+    }
```
You can move the declaration of this function outside this listener.

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,248 @@
+
+    if (event.target.value === 'az') {
+        sortedList.sort(searchByName);
+        renderFriendsList(sortedList);
+
+    } else if (event.target.value === 'za') {
+        sortedList.sort((a, b) => searchByName(b, a));
+        renderFriendsList(sortedList);
+
+    } else if (event.target.value === 'young') {
+        sortedList.sort(sortByAge);
+        renderFriendsList(sortedList);
+
+    } else if (event.target.value === 'older') {
+        sortedList.sort((a, b) => sortByAge(b, a));
+        renderFriendsList(sortedList);
```
`renderFriendsList(sortedList);` is repeated everywhere. DRY

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,248 @@
+    }
+
+    if (event.target.value === 'male') {
+        filteredByGenderList = genderNeutralList.filter(function (friend) {
+            return friend.gender === 'male';
+        })
```
```suggestion
        filteredByGenderList = genderNeutralList.filter(function (friend) {
            return friend.gender === event.target.value;
        })
```

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,248 @@
+    let timer = 0;
+    if (scrollDepth > 100) {
+        window.scrollTo(pageXOffset, scrollDepth);
+        scrollDepth = scrollDepth - 60;
```
Why it is 60 and what is 15? [Magic numbers](https://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad)!

Evgeniy241984 replied:
Resolved

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,248 @@
+const REQUEST_URL = 'https://randomuser.me/api/?results=20',
+    HEADER = document.querySelector('.header-name'),
+    FRIENDS_LIST_WRAP = document.querySelector('.friendsList-wrap'),
+    MENU_WRAP = document.querySelector('.menu-wrap'),
+    SEARCH_FIELD = document.querySelector('#search'),
+    SORTING_MENU = document.querySelector('.sorting-menu'),
+    FILTER_MENU = document.querySelector('.filter-menu'),
+    MOBILE_MENU_BUTTON = document.querySelector('.button-container');
```
Don’t overuse global variables, keep data that you are using closest as possible to the place where you using it. As an example, you have saved elements that were used only in one place. So remove redundant globals

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,248 @@
+    errorButton,
+    genderNeutralList, filteredByGenderList,
+    scrollDepth,
+    lettersArr = HEADER.innerHTML.split('');
```
Ehm... what? What for? Just wrap every letter in span in HTML in a first place.

Evgeniy241984 replied:
i removed this solution

zonzujiro commented:
Please, resolve outdated

zonzujiro commented:
> Please, resolve outdated

:)

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,228 @@
+}
+
+function handleErrors(response) {
+    if (response.status > 400) {
```
```suggestion
    if (response.ok) {
```

Evgeniy241984 replied:
I  used   if ( !response.ok ), cause it is Errors response check
Fixed!

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,228 @@
+function renderFriendsList(friends) {
+    FRIENDS_LIST_WRAP.innerHTML = '';
+
+    friends.forEach(function (friend) {
+        renderFriendsCard(friend);
+    });
```
```suggestion
    friends.forEach(renderFriendsCard);
```

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,228 @@
+
+function renderFriendsCard(friend) {
+    const friendCard = document.createElement('div');
+    friendCard.innerHTML =
+        `
```
Why you need line break here?

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,228 @@
+        })
+
+    } else if (input.value === 'female') {
+        filteredByGenderList = unFilteredList.filter(function (friend) {
+            return friend.gender === input.value;
+        })
```
Code duplication with lines 174-176

Evgeniy241984 replied:
done

zonzujiro commented on submissions/evgeniy24/friends_app/app.js:
```diff
@@ -0,0 +1,228 @@
+
+FILTER_MENU.addEventListener('input', filterFriends);
+
+const menuHiding = function (event) {
```
`hideMenu`

zonzujiro reviewed (approved):
LGTM
